File: short_questions/compny-coding-problems/java/AdditionalStrings.java
package java_solutions;

import java.util.*;

public class AdditionalStrings {

    // 1. Find longest palindrome substring
    public static String longestPalindrome(String s) {
        if (s == null || s.length() < 1)
            return "";
        int start = 0, end = 0;
        for (int i = 0; i < s.length(); i++) {
            int len1 = expandAroundCenter(s, i, i);
            int len2 = expandAroundCenter(s, i, i + 1);
            int len = Math.max(len1, len2);
            if (len > end - start) {
                start = i - (len - 1) / 2;
                end = i + len / 2;
            }
        }
        return s.substring(start, end + 1);
    }

    private static int expandAroundCenter(String s, int left, int right) {
        while (left >= 0 && right < s.length() && s.charAt(left) == s.charAt(right)) {
            left--;
            right++;
        }
        return right - left - 1;
    }

    // 2. Find longest common prefix
    public static String longestCommonPrefix(String[] strs) {
        if (strs == null || strs.length == 0)
            return "";
        String prefix = strs[0];
        for (int i = 1; i < strs.length; i++) {
            while (strs[i].indexOf(prefix) != 0) {
                prefix = prefix.substring(0, prefix.length() - 1);
                if (prefix.isEmpty())
                    return "";
            }
        }
        return prefix;
    }

    // 3. Check if string contains only digits
    public static boolean isDigitsOnly(String str) {
        for (char ch : str.toCharArray()) {
            if (!Character.isDigit(ch))
                return false;
        }
        return true;
    }

    // 4. Count uppercase & lowercase letters
    public static void countCase(String str) {
        int upper = 0, lower = 0;
        for (char ch : str.toCharArray()) {
            if (Character.isUpperCase(ch))
                upper++;
            else if (Character.isLowerCase(ch))
                lower++;
        }
        System.out.println("Upper: " + upper + ", Lower: " + lower);
    }

    // 5. Remove special characters
    public static String removeSpecialChars(String str) {
        return str.replaceAll("[^a-zA-Z0-9]", "");
    }

    // 6. Find all permutations of string (basic)
    public static void permute(String str, int l, int r) {
        if (l == r)
            System.out.println(str);
        else {
            for (int i = l; i <= r; i++) {
                str = swap(str, l, i);
                permute(str, l + 1, r);
                str = swap(str, l, i);
            }
        }
    }

    private static String swap(String str, int i, int j) {
        char[] arr = str.toCharArray();
        char temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
        return new String(arr);
    }

    // 7. Check valid parentheses
    public static boolean isValidParentheses(String s) {
        Stack<Character> stack = new Stack<>();
        for (char c : s.toCharArray()) {
            if (c == '(' || c == '{' || c == '[')
                stack.push(c);
            else {
                if (stack.isEmpty())
                    return false;
                char top = stack.pop();
                if ((c == ')' && top != '(') ||
                        (c == '}' && top != '{') ||
                        (c == ']' && top != '['))
                    return false;
            }
        }
        return stack.isEmpty();
    }

    // 8. Find duplicate words in sentence
    public static void findDuplicateWords(String sentence) {
        String[] words = sentence.toLowerCase().split("\\s+");
        Map<String, Integer> map = new HashMap<>();
        for (String word : words) {
            map.put(word, map.getOrDefault(word, 0) + 1);
        }
        for (Map.Entry<String, Integer> entry : map.entrySet()) {
            if (entry.getValue() > 1) {
                System.out.print(entry.getKey() + " ");
            }
        }
        System.out.println();
    }

    // 9. Reverse each word in place
    public static String reverseEachWord(String sentence) {
        String[] words = sentence.split(" ");
        StringBuilder result = new StringBuilder();
        for (String word : words) {
            result.append(new StringBuilder(word).reverse()).append(" ");
        }
        return result.toString().trim();
    }

    // 10. Check isomorphic strings
    public static boolean isIsomorphic(String s, String t) {
        if (s.length() != t.length())
            return false;
        Map<Character, Character> mapS = new HashMap<>();
        Map<Character, Character> mapT = new HashMap<>();
        for (int i = 0; i < s.length(); i++) {
            char ss = s.charAt(i);
            char tt = t.charAt(i);
            if ((mapS.containsKey(ss) && mapS.get(ss) != tt) ||
                    (mapT.containsKey(tt) && mapT.get(tt) != ss)) {
                return false;
            }
            mapS.put(ss, tt);
            mapT.put(tt, ss);
        }
        return true;
    }

    // 11. Check pangram
    public static boolean isPangram(String str) {
        Set<Character> set = new HashSet<>();
        for (char ch : str.toLowerCase().toCharArray()) {
            if (Character.isLetter(ch)) {
                set.add(ch);
            }
        }
        return set.size() == 26;
    }

    // 12. Print all substrings
    public static void printSubstrings(String str) {
        for (int i = 0; i < str.length(); i++) {
            for (int j = i + 1; j <= str.length(); j++) {
                System.out.println(str.substring(i, j));
            }
        }
    }

    // 13. Remove consecutive duplicates
    public static String removeConsecutiveDeep(String str) {
        if (str.length() < 2)
            return str;
        StringBuilder sb = new StringBuilder();
        sb.append(str.charAt(0));
        for (int i = 1; i < str.length(); i++) {
            if (str.charAt(i) != str.charAt(i - 1)) {
                sb.append(str.charAt(i));
            }
        }
        return sb.toString();
    }

    // 14. Check if strings differ by one character
    public static boolean differByOne(String s1, String s2) {
        if (Math.abs(s1.length() - s2.length()) > 1)
            return false;
        int count = 0;
        int i = 0, j = 0;
        while (i < s1.length() && j < s2.length()) {
            if (s1.charAt(i) != s2.charAt(j)) {
                count++;
                if (count > 1)
                    return false;
                if (s1.length() > s2.length())
                    i++;
                else if (s2.length() > s1.length())
                    j++;
                else {
                    i++;
                    j++;
                }
            } else {
                i++;
                j++;
            }
        }
        if (i < s1.length() || j < s2.length())
            count++;
        return count == 1;
    }

    // 15. Find smallest & largest word
    public static void minMaxWords(String sentence) {
        String[] words = sentence.split("\\s+");
        if (words.length == 0)
            return;
        String min = words[0], max = words[0];
        for (String w : words) {
            if (w.length() < min.length())
                min = w;
            if (w.length() > max.length())
                max = w;
        }
        System.out.println("Min: " + min + ", Max: " + max);
    }

    public static void main(String[] args) {
        System.out.println("1. Longest Palindrome: babad -> " + longestPalindrome("babad"));
        System.out.println(
                "2. Longest Common Prefix: " + longestCommonPrefix(new String[] { "flower", "flow", "flight" }));
        System.out.println("3. Is Digits Only: 12345 -> " + isDigitsOnly("12345"));
        System.out.print("4. Count Case: Hello World -> ");
        countCase("Hello World");
        System.out.println("5. Remove Special Chars: $Hem$lo_World -> " + removeSpecialChars("$Hem$lo_World"));
        System.out.println("6. Permutations: ABC");
        permute("ABC", 0, 2);
        System.out.println("7. Valid Parentheses: ((())) -> " + isValidParentheses("((()))"));
        System.out.print("8. Duplicate Words: ");
        findDuplicateWords("Big black bug bit a big black dog");
        System.out.println("9. Reverse Each Word: Hello World -> " + reverseEachWord("Hello World"));
        System.out.println("10. IsOsomorphic: egg, add -> " + isIsomorphic("egg", "add"));
        System.out.println("11. Is Pangram: The quick brown fox jumps over the lazy dog -> "
                + isPangram("The quick brown fox jumps over the lazy dog"));
        System.out.println("12. Substrings: abc");
        printSubstrings("abc");
        System.out.println("13. Remove Consecutive Dups: aaabbc -> " + removeConsecutiveDeep("aaabbc"));
        System.out.println("14. Differ By One: abc, abd -> " + differByOne("abc", "abd"));
        System.out.print("15. Min Max Words: This is a test string -> ");
        minMaxWords("This is a test string");
    }
}

File: short_questions/compny-coding-problems/golang/additional_strings.go
package main

import (
	"fmt"
	"math"
	"strings"
)

// 1. Find longest palindrome substring
func longestPalindrome(s string) string {
	if len(s) < 1 {
		return ""
	}
	start, end := 0, 0
	for i := 0; i < len(s); i++ {
		len1 := expandAroundCenter(s, i, i)
		len2 := expandAroundCenter(s, i, i+1)
		l := int(math.Max(float64(len1), float64(len2)))
		if l > end-start {
			start = i - (l-1)/2
			end = i + l/2
		}
	}
	return s[start : end+1]
}

func expandAroundCenter(s string, left, right int) int {
	for left >= 0 && right < len(s) && s[left] == s[right] {
		left--
		right++
	}
	return right - left - 1
}

// 2. Find longest common prefix
func longestCommonPrefix(strs []string) string {
	if len(strs) == 0 {
		return ""
	}
	prefix := strs[0]
	for i := 1; i < len(strs); i++ {
		for !strings.HasPrefix(strs[i], prefix) {
			prefix = prefix[:len(prefix)-1]
			if len(prefix) == 0 {
				return ""
			}
		}
	}
	return prefix
}

// 3. Check if string contains only digits
func isDigitsOnly(str string) bool {
	for _, char := range str {
		if char < '0' || char > '9' {
			return false
		}
	}
	return true
}

// 4. Count uppercase & lowercase letters
func countCase(str string) {
	upper, lower := 0, 0
	for _, char := range str {
		if char >= 'A' && char <= 'Z' {
			upper++
		}
		if char >= 'a' && char <= 'z' {
			lower++
		}
	}
	fmt.Printf("Upper: %d, Lower: %d\n", upper, lower)
}

// 5. Remove special characters
func removeSpecialChars(str string) string {
	var result strings.Builder
	for _, char := range str {
		if (char >= 'A' && char <= 'Z') || (char >= 'a' && char <= 'z') || (char >= '0' && char <= '9') {
			result.WriteRune(char)
		}
	}
	return result.String()
}

// 6. Find all permutations of string (basic)
func permuteString(str string, l, r int) {
	if l == r {
		fmt.Println(str)
	} else {
		runes := []rune(str)
		for i := l; i <= r; i++ {
			runes[l], runes[i] = runes[i], runes[l]
			permuteString(string(runes), l+1, r)
			runes[l], runes[i] = runes[i], runes[l] // backtrack
		}
	}
}

// 7. Check valid parentheses
func isValidParentheses(s string) bool {
	var stack []rune
	for _, char := range s {
		if char == '(' || char == '{' || char == '[' {
			stack = append(stack, char)
		} else {
			if len(stack) == 0 {
				return false
			}
			top := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			if (char == ')' && top != '(') ||
				(char == '}' && top != '{') ||
				(char == ']' && top != '[') {
				return false
			}
		}
	}
	return len(stack) == 0
}

// 8. Find duplicate words in string
func findDuplicateWords(sentence string) {
	words := strings.Fields(sentence)
	countMap := make(map[string]int)
	for _, word := range words {
		countMap[strings.ToLower(word)]++
	}
	for word, count := range countMap {
		if count > 1 {
			fmt.Printf("%s ", word)
		}
	}
	fmt.Println()
}

// 9. Count occurrence of given character
func countChar(str string, target rune) int {
	count := 0
	for _, char := range str {
		if char == target {
			count++
		}
	}
	return count
}

// 10. Check if two strings are rotation (Already covered Q19)

// 11. Check pangram
func isPangram(sentence string) bool {
	seen := make(map[rune]bool)
	for _, char := range strings.ToLower(sentence) {
		if char >= 'a' && char <= 'z' {
			seen[char] = true
		}
	}
	return len(seen) == 26
}

// 12. Print all substrings
func printSubstrings(str string) {
	n := len(str)
	for i := 0; i < n; i++ {
		for j := i + 1; j <= n; j++ {
			fmt.Println(str[i:j])
		}
	}
}

// 13. Remove consecutive duplicates
func removeConsecutiveDeep(str string) string {
	if len(str) < 2 {
		return str
	}
	var result strings.Builder
	result.WriteByte(str[0])
	for i := 1; i < len(str); i++ {
		if str[i] != str[i-1] {
			result.WriteByte(str[i])
		}
	}
	return result.String()
}

// 14. Check if strings differ by one character
func differByOne(s1, s2 string) bool {
	if int(math.Abs(float64(len(s1)-len(s2)))) > 1 {
		return false
	}
	countDiff := 0
	i, j := 0, 0
	for i < len(s1) && j < len(s2) {
		if s1[i] != s2[j] {
			countDiff++
			if countDiff > 1 {
				return false
			}
			if len(s1) > len(s2) {
				i++
			} else if len(s2) > len(s1) {
				j++
			} else {
				i++
				j++
			}
		} else {
			i++
			j++
		}
	}
	if i < len(s1) || j < len(s2) {
		countDiff++
	}
	return countDiff == 1
}

// 15. Find smallest & largest word
func minMaxWords(sentence string) {
	words := strings.Fields(sentence)
	if len(words) == 0 {
		return
	}
	minWord, maxWord := words[0], words[0]
	for _, word := range words {
		if len(word) < len(minWord) {
			minWord = word
		}
		if len(word) > len(maxWord) {
			maxWord = word
		}
	}
	fmt.Printf("Min: %s, Max: %s\n", minWord, maxWord)
}

func main() {
	fmt.Println("1. Longest Palindrome:")
	fmt.Println("babad ->", longestPalindrome("babad"))

	fmt.Println("\n2. Longest Common Prefix:")
	fmt.Println("flower, flow, flight ->", longestCommonPrefix([]string{"flower", "flow", "flight"}))

	fmt.Println("\n3. Is Digits Only:")
	fmt.Println("12345 ->", isDigitsOnly("12345"))

	fmt.Println("\n4. Count Case:")
	countCase("Hello World")

	fmt.Println("\n5. Remove Special Chars:")
	fmt.Println("$Hem$lo_World ->", removeSpecialChars("$Hem$lo_World"))

	fmt.Println("\n6. Permutations (ABC):")
	permuteString("ABC", 0, 2)

	fmt.Println("\n7. Valid Parentheses:")
	fmt.Println("((())) ->", isValidParentheses("((()))"))

	fmt.Println("\n8. Duplicate Words:")
	findDuplicateWords("Hello world hello")

	fmt.Println("\n9. Count Char 'l':")
	fmt.Println("hello ->", countChar("hello", 'l'))

	fmt.Println("\n11. Is Pangram:")
	fmt.Println("The quick brown fox jumps over the lazy dog ->", isPangram("The quick brown fox jumps over the lazy dog"))

	fmt.Println("\n12. Substrings (abc):")
	printSubstrings("abc")

	fmt.Println("\n13. Remove Consecutive Dups:")
	fmt.Println("aaabbc ->", removeConsecutiveDeep("aaabbc"))

	fmt.Println("\n14. Differ By One:")
	fmt.Println("abc, abd ->", differByOne("abc", "abd"))

	fmt.Println("\n15. Min Max Words:")
	minMaxWords("This is a test string")
}
